### ISSUE TITLE
Add "TabNet" model to regression and classification algorithms

### ISSUE THREAD

**Issue by tsereno** [pycaret/pycaret#1413]
**Is your feature request related to a problem? Please describe.**
Opportunity to improve model accuracy with ability to ensemble TabNet model (deep learning) with existing set of models. 

**Describe the solution you'd like**
Add native support for TabNet.

**Describe alternatives you've considered**
Considered doing the ensemble using other libraries.

**Additional context**
There is a pytorch version of TabNet available now which integrates in with sci-kit learn, https://pythonrepo.com/repo/dreamquark-ai-tabnet-python-pytorch-utilities

**Comment by atwahsz** [comment]
hello @tsereno , this looks interesting , lets work on it together and provide the implementation to the team

**Comment by pycaret** [comment]
This is a great idea. I am in full support of this. 

Are you interested working on this?

We will provide whatever support we can. 

Cc: @Yard1

**Comment by atwahsz** [comment]
yes totally interested , if possible can we pass a variable for deep learning , where the split will be train valid and test . so each fold can run normally

**Comment by tsereno** [comment]
Yes, the fit function for TabNet has parameter for train dataset and validation dataset (eval_set parameter), same as Scikit learn fit function.


Sent from Yahoo Mail for iPhone


On Saturday, July 3, 2021, 7:10 AM, Saleh Z. Atwah ***@***.***> wrote:




yes totally interested , if possible can we pass a variable for deep learning , where the split will be train valid and test . so each fold can run normally

**Comment by IncubatorShokuhou** [comment]
Hi @tsereno , I just write a simple demo using tabnet by `create_model`, but I am wondering how can we integrete DL models into PyCaret as I don't know how to properly turn a DL model inside under the current framework, which might be quite different from current models ( including batchsize, learning rate scheduler, optimizer et. al. , and other more detailed network structure).
Here is my demo FYI. Hope you can provide constructive feedback, especially in tuning models.
```python
from pycaret.datasets import get_data
from pytorch_tabnet.tab_model import TabNetRegressor
from pycaret.regression import *
from sklearn.base import BaseEstimator
from sklearn.utils import check_array
import numpy as np

data = get_data("diamond")
exp_reg101 = setup(data=data, target="Price", session_id=123, silent=True, n_jobs=-1)


class tabnet(BaseEstimator):
    # create a class with scikit-learn compatible API
    # as `TabNetRegressor().fit` only receives y.shape = (n,1) instead of (n,) as input
    def __init__(self):
        self.tabnet = TabNetRegressor()

    def fit(self, X, y):
        return self.tabnet.fit(check_array(X), np.array(y).reshape(-1, 1))

    def predict(self, X):
        return self.tabnet.predict(check_array(X))


tabnet_reg = create_model(tabnet())

compare_models(include=[tabnet_reg, "lightgbm", "rf"])

```

**Comment by IncubatorShokuhou** [comment]
What's more, TabNet uses embedding, instead of one-hot encoding et. al. used in Pycaret, to handle category features, which is completely different from PyCaret. As far as I am concerned, most of the CTR models based on DL also uses embedding, which means if we are trying to put DL models into PyCaret, we may need to introduce huge changes, especially in preprocessing. @pycaret I think we may need consider more deep in this issue.

**Comment by tsereno** [comment]
I follow what you are saying about embeddings and possibly introducing big changes into the pipeline. Possibly a better approach would be to directly integrate tensorflow and pytorch into the pipeline first to directly support DL and then add in state of the art DL algorithms for each module (classification, regression, etc).
However, I keep thinking that MLP is a neural network and already works great with the current preprocessing (such as the one-hot encoding) in PyCaret.  MLP is considered a subset of a DL/DNN.  I need to read some more papers on TabNet, but it would seem that same preprocessed dataset could be used to create a good model.    On Wednesday, July 14, 2021, 07:51:02 PM PDT, Hao Lyu ***@***.***> wrote:  
 
 


What's more, TabNet uses embedding, instead of one-hot encoding et. al. used in Pycaret, to handle category features, which is completely different from PyCaret. As far as I am concerned, most of the CTR models based on DL also uses embedding, which means if we are trying to put DL models into PyCaret, we may need to introduce huge changes, especially in preprocessing. @pycaret I think we may need consider more deep in this issue.

**Comment by IncubatorShokuhou** [comment]
> directly integrate tensorflow and pytorch into the pipeline first to directly support DL

Pytorch has been able to be integrated to PyCaret (using [skorch](https://github.com/skorch-dev/skorch) ). Here is a demo written by me: https://github.com/pycaret/pycaret/issues/700#issuecomment-879700610 

However, if we are looking forward to a deep and complete integration, there remains a lot of things that we need to do. For examples, ddp for Pytorch model, completely diffreent preprocessing methods (i.e. embedding mentioned above) from regular ml model, potential detailed changes inside the DL model, complex training methods (lr scheduler, optimizer, metric) et. al. ,which may means a totally different pipeline. And all the thing mentioned above seems to quite hard to resolve in current stage.

Also, I believe that a DL model should be more customized, and people are unable to get a satisfactory method if he/she does not know about the details of his/her DL model, and implements the model, including training, tuning et. al step by step on his/her own.

So as a conclusion, I am not fully support that a DL model is easily integrated into PyCaret at this stage.

**Comment by tsereno** [comment]
Thank you for putting together this demo code.  I will attempt to tune hyper-parameters using the diamond dataset and revert back with results.    On Wednesday, July 14, 2021, 07:32:52 PM PDT, Hao Lyu ***@***.***> wrote:  
 
 


Hi @tsereno , I just write a simple demo using tabnet by create_model, but I am wondering how can we integrete DL models into PyCaret as I don't know how to properly turn a DL model inside under the current framework, which might be quite different from current models ( including batchsize, learning rate scheduler, optimizer et. al. , and other more detailed network structure).
Here is my demo FYI. Hope you can provide constructive feedback, especially in tuning models.
from pycaret.datasets import get_data
from pytorch_tabnet.tab_model import TabNetRegressor
from pycaret.regression import *
from sklearn.base import BaseEstimator
from sklearn.utils import check_array
import numpy as np

data = get_data("diamond")
exp_reg101 = setup(data=data, target="Price", session_id=123, silent=True, n_jobs=-1)


class tabnet(BaseEstimator):
    # create a class with scikit-learn compatible API
    # as `TabNetRegressor().fit` only receives y.shape = (n,1) instead of (n,) as input
    def __init__(self):
        self.tabnet = TabNetRegressor()

    def fit(self, X, y):
        return self.tabnet.fit(check_array(X), np.array(y).reshape(-1, 1))

    def predict(self, X):
        return self.tabnet.predict(check_array(X))


tabnet_reg = create_model(tabnet())

compare_models(include=[tabnet_reg, "lightgbm", "rf"])

**Comment by IncubatorShokuhou** [comment]
> Thank you for putting together this demo code.  I will attempt to tune hyper-parameters using the diamond dataset and revert back with results. 
I would like to suggest you not to use any preprocessing method in Pycaret because by default PyCaret handle the catalogical feature with one-hot, which means all the `X_train` will be numerical, which is terrible for using a embedding DL model.

**Comment by IncubatorShokuhou** [comment]
> > Thank you for putting together this demo code.  I will attempt to tune hyper-parameters using the diamond dataset and revert back with results.

I would like to suggest you not to use any preprocessing method in Pycaret because by default PyCaret handle the catalogical feature with one-hot, which means all the `X_train` will be numerical, which is terrible for using a embedding DL model.

**Comment by tsereno** [comment]
That makes sense, I will try without pre-processing when I get a chance.
With pre-processing, I used parameters as follows.return self.tabnet.fit(check_array(X), np.array(y).reshape(-1, 1), max_epochs=1000, patience=3, batch_size=2048, virtual_batch_size=512)
The model comparison selected tabnet as the best model.
compare_models(include=[tabnet_reg, "lightgbm", "rf"],cross_validation=False)
Processing:
|  |  |  |
|  |  |  |
| Initiated | . . . . . . . . . . . . . . . . . . | 08:13:44 |
| Status | . . . . . . . . . . . . . . . . . . | Compiling Final Models |
| Estimator | . . . . . . . . . . . . . . . . . . | tabnet |


|  | Model | MAE | MSE | RMSE | R2 | RMSLE | MAPE | TT (Sec) |
| 0 | tabnet | 744.9703 | 1.881065e+06 | 1371.5193 | 0.9820 | 0.0820 | 0.0619 | 98.91 |
| 2 | Random Forest Regressor | 747.6673 | 2.673768e+06 | 1635.1659 | 0.9744 | 0.0756 | 0.0566 | 1.35 |
| 1 | Light Gradient Boosting Machine | 755.9333 | 3.111456e+06 | 1763.9319 | 0.9702 | 0.0758 | 0.0558 | 0.07 |

tabnet()
    On Thursday, July 15, 2021, 12:32:23 AM PDT, Hao Lyu ***@***.***> wrote:  
 
 



Thank you for putting together this demo code.  I will attempt to tune hyper-parameters using the diamond dataset and revert back with results.
I would like to suggest you not to use any preprocessing method in Pycaret because by default PyCaret handle the catalogical feature with one-hot, which means all the X_train will be numerical, which is terrible for using a embedding DL model.

**Comment by tsereno** [comment]
Tabnet doesn't train without pre-processing if there are any string columns, because the strings need to be embedded.
So, I used same embedding process used in the following notebook example dreamquark-ai/tabnet
Only a few tweaks to make it work for the "diamond" dataset and used same parameters for training.  Metrics look ok.
clf.fit(    X_train=X_train, y_train=y_train,    eval_set=[(X_train, y_train), (X_valid, y_valid)],    eval_name=['train', 'valid'],    eval_metric=['rmsle', 'mae', 'rmse', 'mse'],    max_epochs=max_epochs,    patience=50,    batch_size=2048, virtual_batch_size=512,    num_workers=2,    drop_last=False) ...epoch 277| loss: 3174291.62857| train_rmsle: 0.01239 | train_mae: 741.71867| train_rmse: 1050.40308| train_mse: 1103346.63725| valid_rmsle: 0.01254 | valid_mae: 905.79997| valid_rmse: 1418.90392| valid_mse: 2013288.33737| 0:03:32s
...Early stopping occurred at epoch 327 with best_epoch = 277 and best_valid_mse = 2013288.33737Best weights from best epoch are automatically used!
    On Thursday, July 15, 2021, 01:24:20 AM PDT, Tim Sereno ***@***.***> wrote:  
 
  That makes sense, I will try without pre-processing when I get a chance.
With pre-processing, I used parameters as follows.return self.tabnet.fit(check_array(X), np.array(y).reshape(-1, 1), max_epochs=1000, patience=3, batch_size=2048, virtual_batch_size=512)
The model comparison selected tabnet as the best model.
compare_models(include=[tabnet_reg, "lightgbm", "rf"],cross_validation=False)
Processing:
|  |  |  |
|  |  |  |
| Initiated | . . . . . . . . . . . . . . . . . . | 08:13:44 |
| Status | . . . . . . . . . . . . . . . . . . | Compiling Final Models |
| Estimator | . . . . . . . . . . . . . . . . . . | tabnet |


|  | Model | MAE | MSE | RMSE | R2 | RMSLE | MAPE | TT (Sec) |
| 0 | tabnet | 744.9703 | 1.881065e+06 | 1371.5193 | 0.9820 | 0.0820 | 0.0619 | 98.91 |
| 2 | Random Forest Regressor | 747.6673 | 2.673768e+06 | 1635.1659 | 0.9744 | 0.0756 | 0.0566 | 1.35 |
| 1 | Light Gradient Boosting Machine | 755.9333 | 3.111456e+06 | 1763.9319 | 0.9702 | 0.0758 | 0.0558 | 0.07 |

tabnet()
    On Thursday, July 15, 2021, 12:32:23 AM PDT, Hao Lyu ***@***.***> wrote:  
 
 



Thank you for putting together this demo code.  I will attempt to tune hyper-parameters using the diamond dataset and revert back with results.
I would like to suggest you not to use any preprocessing method in Pycaret because by default PyCaret handle the catalogical feature with one-hot, which means all the X_train will be numerical, which is terrible for using a embedding DL model.

**Comment by nicholasdomene** [comment]
@tsereno Are you still working on this issue? If so, I'd be willing to help.

**Comment by tsereno** [comment]
I was able to use tabnet with pycaret using the demo code provided (BaseEstimator), and using the encoding method required for tabnet.  However, I wasn't able to do ensemble, stacking, bagging, etc. on a custom model (I tried with other algorithms such as NGBoost as well) in pycaret as it returned errors.  Since then there was an issue resolution for doing ensemble of custom model, but it still returned errors and I could not get it to work.  Also, I could not get tabnet model provide better metrics than xgboost and lightgbm even with lots of manual hyperparameter tuning which is painfully slow.
I took a different direction since then and started using AutoGluon python library which has satisfied my needs.  It has simple automl parameters to ensemble/bag/stack multiple algorithms including xgboost, lightgbm, catboost, and deep neural network from fastapi including simple parameters to do the automatic hyperparameter tuning (i.e. bayesian) with no custom code required.  Pre-processing is automatic which eliminates custom code to do the encoding.  However, I still haven't run across a dataset yet where the deep neural network model beats the other algorithms for regression and training deep learning model is painfully slow even on small datasets.


    On Saturday, November 6, 2021, 05:42:24 PM PDT, nicholasdomene ***@***.***> wrote:  
 
 


@tsereno Are you still working on this issue? If so, I'd be willing to help.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by moezali1** [comment]
Closing as part of the 4.0 backlog cleanup — see #4177.

PyCaret 4.0 is a ground-up rewrite (OOP-only API, sklearn-native engine, Python 3.11+, sklearn ≥ 1.7) and almost all pre-4.0 issues either reference removed APIs (`setup()`, `plot_model`, `evaluate_model`, `interpret_model`, `automl`, the legacy time-series module, the Streamlit dashboard, …) or no longer reproduce on the new code path.

**If this still affects you in 4.0, please re-file as a new issue** with:

- the experiment constructor call (`Experiment(...)`),
- the operation that fails,
- the full traceback,
- your Python / sklearn / pycaret versions.

Resources for 4.0:

- Docs: https://pycaret.org/docs
- Migrate from 3.x: https://pycaret.org/docs/guides/migrating-from-3-x
- API reference: https://pycaret.org/reference
- FAQ: https://pycaret.org/docs/resources/faqs

Thanks for filing — and apologies for the long latency. The repo had drifted; we're starting clean.
